cpp/src/io/utilities/data_casting.cu

```diff
@@ -0,0 +1,928 @@
+      };
+      state_table curr{is_within_bounds && c == '\\', false};  // state transition vector.
+      auto composite_op = [](state_table op1, state_table op2) {
+        return state_table{op2.state[op1.state[0]], op2.state[op1.state[1]]};
+      };
+      state_table scanned;
+      // inclusive scan of escaping backslashes
+      if constexpr (is_warp) {
+        using SlashScan = cub::WarpScan<state_table>;
+        __shared__ typename SlashScan::TempStorage temp_slash[num_warps];
+        SlashScan(temp_slash[warp_id]).InclusiveScan(curr, scanned, composite_op);
+        is_escaping_backslash = scanned.state[init_state];
+        init_state            = __shfl_sync(MASK, is_escaping_backslash, BLOCK_SIZE - 1);
+        __syncwarp();
+        is_slash.shift(warp_id);
+        is_slash.set_bits(warp_id, is_escaping_backslash);
+        is_prev_escaping_backslash = is_slash.get_bit(warp_id, lane - 1);
+      } else {
+        using SlashScan = cub::BlockScan<state_table, BLOCK_SIZE>;
+        __shared__ typename SlashScan::TempStorage temp_slash;
+        SlashScan(temp_slash).InclusiveScan(curr, scanned, composite_op);
+        is_escaping_backslash = scanned.state[init_state];
+        __syncthreads();
+        if (threadIdx.x == BLOCK_SIZE - 1) init_state = is_escaping_backslash;
+        __syncthreads();
+        is_slash.shift(warp_id);
+        is_slash.set_bits(warp_id, is_escaping_backslash);
+        is_prev_escaping_backslash = is_slash.get_bit(warp_id, lane - 1);
+        // There is another __syncthreads() at the end of for-loop.
+      }
+
+      bool error = false;
+      if (is_within_bounds) {
+        // instead of '\\', using is_escaping_backslash, and previous index value also here.
+        error |= (is_escaping_backslash /*c == '\\'*/ && char_index == (in_end - in_begin) - 1);
+        error |= (is_prev_escaping_backslash && escaped_char == NON_ESCAPE_CHAR);
+        error |= (is_prev_escaping_backslash && c == 'u' &&
+                  ((in_begin + char_index + UNICODE_HEX_DIGIT_COUNT >= in_end) |
+                   !is_hex(in_begin[char_index + 1]) | !is_hex(in_begin[char_index + 2]) |
+                   !is_hex(in_begin[char_index + 3]) | !is_hex(in_begin[char_index + 4])));
+      }
+      // propagate error using warp shuffle.
+      if constexpr (is_warp) {
+        error = __any_sync(MASK, error);
+      } else {
+        using ErrorReduce = cub::BlockReduce<bool, BLOCK_SIZE>;
+        __shared__ typename ErrorReduce::TempStorage temp_storage_error;
+        __shared__ bool error_reduced;
+        error_reduced = ErrorReduce(temp_storage_error).Sum(error);  // TODO use cub::LogicalOR.
+        // only valid in thread0, so shared memory is used for broadcast.
+        __syncthreads();
+        error = error_reduced;
+      }
+      if (error) {
+        if (lane == 0) {
+          if (null_mask != nullptr) {
+            clear_bit(null_mask, istring);
+            atomicAdd(null_count_data, 1);
```

Comment:
question: Do we need to make sure that `clear_bit` and especially `atomicAdd` is only executed if `!d_chars`? Wouldn't it we count the null twice otherwise?

Reply:
oh. that didn't happen because nulls are skipped in line 442-444
size computation kernel call sets the null, writing output kernel call skips it early.